File: src/nn_model.rs
// code_files/src/nn_model.rs
use tch::{nn, Tensor};
use crate::game_env::{
    ACTION_SPACE_SIZE, BOARD_CHANNELS, BOARD_COLS, BOARD_ROWS,
    SCALAR_FEATURE_COUNT, STATE_STACK_SIZE,
};

// ========== 4×8 棋盘配置 (使用 game_env.rs 的常量) ==========
// 这里的常量定义用于适配 tch 的 tensor 维度需求
// 数值直接衍生自 game_env.rs 中定义的物理环境参数
const TOTAL_CHANNELS: i64 = (BOARD_CHANNELS * STATE_STACK_SIZE) as i64;
const BOARD_H: i64 = BOARD_ROWS as i64;
const BOARD_W: i64 = BOARD_COLS as i64;
const SCALAR_FEATURES: i64 = SCALAR_FEATURE_COUNT as i64;
const ACTION_SIZE: i64 = ACTION_SPACE_SIZE as i64;

// 网络隐藏层通道数
// 针对暗棋的复杂性（尤其是炮的跳吃和翻棋的不确定性），
// 使用较宽的通道数以捕捉更丰富的空间特征
const HIDDEN_CHANNELS: i64 = 128;

// 标准残差块 (Residual Block)
// 结构: Conv -> BN -> ReLU -> Conv -> BN -> (+Input) -> ReLU
// 保持输入输出的空间尺寸和通道数不变，用于构建深层网络而不丢失梯度
struct BasicBlock {
    conv1: nn::Conv2D,
    bn1: nn::BatchNorm,
    conv2: nn::Conv2D,
    bn2: nn::BatchNorm,
}

impl BasicBlock {
    fn new(vs: &nn::Path, channels: i64) -> Self {
        // 使用 padding=1 配合 kernel_size=3 以保持空间尺寸不变
        let conv_cfg = nn::ConvConfig {
            padding: 1,
            ..Default::default()
        };
        Self {
            conv1: nn::conv2d(vs / "conv1", channels, channels, 3, conv_cfg),
            bn1: nn::batch_norm2d(vs / "bn1", channels, Default::default()),
            conv2: nn::conv2d(vs / "conv2", channels, channels, 3, conv_cfg),
            bn2: nn::batch_norm2d(vs / "bn2", channels, Default::default()),
        }
    }

    fn forward_t(&self, xs: &Tensor, train: bool) -> Tensor {
        let residual = xs;
        let out = xs
            .apply(&self.conv1)
            .apply_t(&self.bn1, train)
            .relu()
            .apply(&self.conv2)
            .apply_t(&self.bn2, train);

        // 残差连接：将输入直接加到输出上，然后进行激活
        (out + residual).relu()
    }
}

// BanqiNet: 暗棋策略价值网络
// 架构:
// 1. 输入卷积层 (Input Conv)
// 2. 残差塔 (Residual Tower): 堆叠多个残差块
// 3. 策略头 (Policy Head): 输出所有合法动作的概率 logits
// 4. 价值头 (Value Head): 输出当前局面的胜率评估 [-1, 1]
pub struct BanqiNet {
    // 初始卷积层：将输入状态映射到隐藏特征空间
    conv_input: nn::Conv2D,
    bn_input: nn::BatchNorm,

    // 残差塔：用于提取深层空间特征
    res_block1: BasicBlock,
    res_block2: BasicBlock,
    res_block3: BasicBlock,

    // 全连接层：处理扁平化后的空间特征与全局标量特征
    fc1: nn::Linear,
    fc2: nn::Linear,

    // 输出头
    policy_head: nn::Linear,
    value_head: nn::Linear,
}

impl BanqiNet {
    pub fn new(vs: &nn::Path) -> Self {
        let conv_cfg = nn::ConvConfig {
            padding: 1,
            ..Default::default()
        };

        // 1. Input Conv
        // 将棋盘状态张量转换为隐藏层特征
        let conv_input = nn::conv2d(
            vs / "conv_input",
            TOTAL_CHANNELS,
            HIDDEN_CHANNELS,
            3,
            conv_cfg,
        );
        let bn_input = nn::batch_norm2d(vs / "bn_input", HIDDEN_CHANNELS, Default::default());

        // 2. 残差塔
        // 连续经过3个残差块，深化特征表示
        let res_block1 = BasicBlock::new(&(vs / "res1"), HIDDEN_CHANNELS);
        let res_block2 = BasicBlock::new(&(vs / "res2"), HIDDEN_CHANNELS);
        let res_block3 = BasicBlock::new(&(vs / "res3"), HIDDEN_CHANNELS);

        // 3. 计算全连接层输入维度
        // 空间特征展平大小
        let flat_size = HIDDEN_CHANNELS * BOARD_H * BOARD_W;
        
        // 总输入 = 空间特征 + 全局标量特征
        let total_fc_input = flat_size + SCALAR_FEATURES;

        // 4. 全连接层 (Shared Dense Layers)
        // 融合空间信息和全局信息（如步数、血量、存活棋子统计等）
        let fc1 = nn::linear(vs / "fc1", total_fc_input, 512, Default::default());
        let fc2 = nn::linear(vs / "fc2", 512, 256, Default::default());

        // 5. 输出头
        // Policy Head: 映射到动作空间大小
        // Value Head: 映射到标量价值
        let policy_head = nn::linear(vs / "policy", 256, ACTION_SIZE, Default::default());
        let value_head = nn::linear(vs / "value", 256, 1, Default::default());

        Self {
            conv_input,
            bn_input,
            res_block1,
            res_block2,
            res_block3,
            fc1,
            fc2,
            policy_head,
            value_head,
        }
    }

    fn forward_t(&self, board: &Tensor, scalars: &Tensor, train: bool) -> (Tensor, Tensor) {
        // 1. 输入卷积处理
        // 提取初步的空间特征
        let x = board
            .apply(&self.conv_input)
            .apply_t(&self.bn_input, train)
            .relu();

        // 2. 通过残差塔
        // 深层特征提取
        let x = self.res_block1.forward_t(&x, train);
        let x = self.res_block2.forward_t(&x, train);
        let x = self.res_block3.forward_t(&x, train);

        // 3. 特征展平 (Flatten)
        // 将 [Batch, Channels, H, W] 展平为 [Batch, Features]
        let x = x.flatten(1, -1);

        // 4. 特征拼接 (Concatenate)
        // 将卷积提取的图像特征与全局标量特征拼接
        let combined = Tensor::cat(&[&x, scalars], 1);

        // 5. 全连接层处理
        // 共享的稠密层，用于后续分头
        let shared = combined.apply(&self.fc1).relu().apply(&self.fc2).relu();

        // 6. 输出计算
        // Policy: 输出 Logits (未经过 Softmax)
        // Value: 输出 Tanh 激活的价值 [-1, 1]
        let policy_logits = shared.apply(&self.policy_head);
        let value = shared.apply(&self.value_head).tanh();

        (policy_logits, value)
    }

    /// 训练模式的前向传播
    /// 启用 BatchNorm 的统计更新和 Dropout (如果有)
    pub fn forward(&self, board: &Tensor, scalars: &Tensor) -> (Tensor, Tensor) {
        self.forward_t(board, scalars, true)
    }

    /// 推理模式的前向传播
    /// 使用固定的 BatchNorm 统计量，禁用 Dropout
    pub fn forward_inference(&self, board: &Tensor, scalars: &Tensor) -> (Tensor, Tensor) {
        self.forward_t(board, scalars, false)
    }
}

File: python/nn_model.py
# nn_model.py
import torch
import torch.nn as nn
import torch.nn.functional as F
from constant import (
    TOTAL_INPUT_CHANNELS,
    HIDDEN_CHANNELS,
    NUM_RES_BLOCKS,
    BOARD_ROWS,
    BOARD_COLS,
    SCALAR_FEATURE_COUNT,
    ACTION_SPACE_SIZE
)

class BasicBlock(nn.Module):
    """
    标准残差块，参考 nn_model.rs
    结构: Conv -> BN -> ReLU -> Conv -> BN -> (+Input) -> ReLU
    """
    def __init__(self, channels):
        super(BasicBlock, self).__init__()
        # Rust: padding: 1, kernel_size: 3
        self.conv1 = nn.Conv2d(channels, channels, kernel_size=3, padding=1, bias=False)
        self.bn1 = nn.BatchNorm2d(channels)
        self.conv2 = nn.Conv2d(channels, channels, kernel_size=3, padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(channels)

    def forward(self, x):
        residual = x
        
        out = self.conv1(x)
        out = self.bn1(out)
        out = F.relu(out)
        
        out = self.conv2(out)
        out = self.bn2(out)
        
        # 残差连接
        out += residual
        out = F.relu(out)
        
        return out

class BanqiNet(nn.Module):
    """
    AlphaZero-style policy-value network for 4x8 Dark Chess.

    Input:
      board   - (N, 16, 4, 8) float32
      scalars - (N, 35) float32 (no action_mask — masks handled at loss level)

    Architecture:
      Input conv(16→64, 3×3)
      6 × ResidualBlock (64 ch)
      Policy head: 1×1 conv(64→4) → flatten(128) → +scalars(35) → FC1(163→512) → FC2(512→352)
      Value head:  1×1 conv(64→4) → flatten(128) → +scalars(35) → FC1(163→256) → FC2(256→1) → tanh

    Total params ~1.03M (down from ~2.94M).
    """
    def __init__(self, num_res_blocks=NUM_RES_BLOCKS, hidden_channels=HIDDEN_CHANNELS,
                 policy_channels=4, value_channels=4):
        super(BanqiNet, self).__init__()
        
        # 1. 输入卷积
        self.conv_input = nn.Conv2d(
            TOTAL_INPUT_CHANNELS,
            hidden_channels,
            kernel_size=3,
            padding=1,
            bias=False
        )
        self.bn_input = nn.BatchNorm2d(hidden_channels)
        
        # 2. 残差塔
        self.res_tower = nn.ModuleList(
            [BasicBlock(hidden_channels) for _ in range(num_res_blocks)]
        )

        # 3. 策略头
        self.policy_channels = policy_channels
        self.policy_conv = nn.Conv2d(hidden_channels, policy_channels, kernel_size=1, bias=False)
        self.policy_bn = nn.BatchNorm2d(policy_channels)
        self.policy_flat_size = policy_channels * BOARD_ROWS * BOARD_COLS
        self.policy_fc_input = self.policy_flat_size + SCALAR_FEATURE_COUNT
        self.policy_fc1 = nn.Linear(self.policy_fc_input, 512)
        self.policy_fc2 = nn.Linear(512, ACTION_SPACE_SIZE)

        # 4. 价值头
        self.value_channels = value_channels
        self.value_conv = nn.Conv2d(hidden_channels, value_channels, kernel_size=1, bias=False)
        self.value_bn = nn.BatchNorm2d(value_channels)
        self.value_flat_size = value_channels * BOARD_ROWS * BOARD_COLS
        self.value_fc_input = self.value_flat_size + SCALAR_FEATURE_COUNT
        self.value_fc1 = nn.Linear(self.value_fc_input, 256)
        self.value_fc2 = nn.Linear(256, 1)

    def forward(self, board, scalars):
        """
        Args:
            board: Tensor，形状 (Batch, Channels, H, W)
            scalars: Tensor，形状 (Batch, Scalar_Features)
        Returns:
            policy_logits: (Batch, Action_Size)
            value: (Batch, 1) - Tanh 激活
        """
        # 1. 输入卷积
        x = self.conv_input(board)
        x = self.bn_input(x)
        x = F.relu(x)
        
        # 2. 残差塔
        for block in self.res_tower:
            x = block(x)

        # 3. 策略头
        p = self.policy_conv(x)
        p = self.policy_bn(p)
        p = F.relu(p)
        p = p.view(p.size(0), -1)
        p_combined = torch.cat([p, scalars], dim=1)
        p_out = F.relu(self.policy_fc1(p_combined))
        policy_logits = self.policy_fc2(p_out)

        # 4. 价值头
        v = self.value_conv(x)
        v = self.value_bn(v)
        v = F.relu(v)
        v = v.view(v.size(0), -1)
        v_combined = torch.cat([v, scalars], dim=1)
        v_out = F.relu(self.value_fc1(v_combined))
        value = torch.tanh(self.value_fc2(v_out))
        
        return policy_logits, value

def load_model_weights(model: "nn.Module", path: str, device: "torch.device") -> None:
    """
    把 path 指向的权重文件加载进 model。自动识别三种格式：
      1. TorchScript 归档 (.pt, training_service 中 torch.jit.trace 产出) —— torch.jit.load
      2. 普通 state_dict (.pth)
      3. 完整 checkpoint dict（含 model_state_dict / optimizer_state_dict 等）

    注意：PyTorch 2.6+ 中 torch.load(weights_only=True) 无法加载 TorchScript
    归档（抛 "Cannot use weights_only=True with TorchScript archives"），
    因此必须先尝试 torch.jit.load。全部格式都失败时抛出最后一个异常，
    由调用方决定如何处理（如保留旧模型）。
    """
    try:
        jit_model = torch.jit.load(path, map_location=device)
        model.load_state_dict(jit_model.state_dict())
        return
    except Exception:
        pass
    try:
        state = torch.load(path, map_location=device, weights_only=True)
    except Exception:
        state = torch.load(path, map_location=device)
    if hasattr(state, "state_dict"):
        model.load_state_dict(state.state_dict())
    elif isinstance(state, dict) and "model_state_dict" in state:
        model.load_state_dict(state["model_state_dict"])
    else:
        model.load_state_dict(state)


if __name__ == "__main__":
    # 简单测试：验证维度是否匹配
    batch_size = 4
    dummy_board = torch.randn(batch_size, TOTAL_INPUT_CHANNELS, BOARD_ROWS, BOARD_COLS)
    dummy_scalars = torch.randn(batch_size, SCALAR_FEATURE_COUNT)
    
    model = BanqiNet()
    policy, value = model(dummy_board, dummy_scalars)
    
    print(f"Model Structure: {model}")
    print(f"Input Board: {dummy_board.shape}")
    print(f"Input Scalars: {dummy_scalars.shape}")
    print(f"Output Policy: {policy.shape} (Expected: [{batch_size}, {ACTION_SPACE_SIZE}])")
    print(f"Output Value: {value.shape} (Expected: [{batch_size}, 1])")